include/stream_list_priv.h

```diff
@@ -0,0 +1,123 @@
+ *   from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef SRTP_STREAM_LIST_PRIV_H
+#define SRTP_STREAM_LIST_PRIV_H
+
+#include <srtp_priv.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * srtp_stream_list_t holds a list of srtp_stream_t, each identified
+ * by their SSRC.
+ *
+ * the API was extracted to allow downstreams to override its
+ * implementation by defining the `SRTP_NO_STREAM_LIST` preprocessor
+ * directive, which removes the default implementation of these
+ * functions. if this is done, the `next` & `prev` fields are free for
+ * the implementation to use.
+ *
+ * this is still an internal interface; there is no stability
+ * guarantee--downstreams should watch this file for changes in
+ * signatures or semantics.
+ */
+
+/**
+ * allocate and initialize a stream list instance
+ */
+srtp_err_status_t srtp_stream_list_alloc(srtp_stream_list_t *list_ptr);
+
+/**
+ * deallocate a stream list instance
+ *
+ * the list must be empty or else an error is returned.
+ */
+srtp_err_status_t srtp_stream_list_dealloc(srtp_stream_list_t list);
+
+/**
+ * insert a stream into the list
+ *
+ * returns srtp_err_status_alloc_fail if insertion failed due to unavailable
+ * capacity in the list. if operation succeeds, srtp_err_status_ok is returned
+ * and ownership is transferred from caller into the list
+ *
+ * if another stream with the same SSRC already exists in the list,
+ * behavior is undefined. if the SSRC field is mutated while the
+ * stream is inserted, further operations have undefined behavior
+ */
+srtp_err_status_t srtp_stream_list_insert(srtp_stream_list_t list,
+                                          srtp_stream_t stream);
+
+/*
+ * look up the stream corresponding to the specified SSRC and return it.
+ * if no such SSRC is found, NULL is returned.
+ */
+srtp_stream_t srtp_stream_list_get(srtp_stream_list_t list, uint32_t ssrc);
+
+/**
+ * remove the stream from the list.
+ *
+ * ownership is transferred to the caller.
+ *
+ * if the stream is not in the list the behavior is undefined.
+ */
+void srtp_stream_list_remove(srtp_stream_list_t list, srtp_stream_t stream);
```

Comment:
Considering this method on its own, I would have expected something more like:

```
srtp_stream_t srtp_stream_list_remove(srtp_stream_list_t list, unsigned int ssrc);
```

The API here only really makes sense in combination with `srtp_stream_list_get`, i.e., if the stream pointer came from that method. (It basically *is* that method except for the change of ownership.) We should either document that fact, or make the API more like the above.  My preference would be for the latter, maybe calling it `extract` in parallel with [the comparable C++ STL design pattern](https://en.cppreference.com/w/cpp/container/map/extract).

Reply:
That is what it was before my change, but it is also used inside of a srtp_stream_list_for_each loop in which case the caller already has  the srtp_stream_t so this was to prevent a second lookup. This kind of goes back to what is this PR trying to achieve, minimal change or complete separation.

Reply:
I'm fine with not changing the interface.  I'll suggest some documentation above.